### PR TITLE
[RFC] Incsub 4 (renamed to 'inccommand')

### PR DESCRIPTION
**Pull request by justinmk**
- [x] rename to 'inccommand'
- [x] back out special cases in `do_ecmd()` and `buflist_new()`
- [x] reduce shared/global state
- [x] preserve `b:changedtick` https://github.com/neovim/neovim/pull/5561#discussion_r86656183
- [x] add tests for `feedkeys(':%s/foo/bar/')`, `feedkeys(":%s/foo/bar/\<CR>")`, ... 
    - [x] check `KeyTyped` to avoid preview-mode during `feedkeys()`
- [x] should preview split appear if `:{range}` is in the viewport? (answer: yes, be predictable)
- [x] disable 'cursorline', 'spell' in preview window
- [x] set buffer name to `[Preview]` instead of `[inc_sub]`
- [x] change `[lnum]` display to `|lnum|` (like quickfix window)
- [x] highlight/multiline bug #7231 https://github.com/neovim/neovim/pull/5226#issuecomment-257153602 
- [ ] re-use _window id_ for preview buffer (`:h win_getid()`)
    - best approach here would be to _keep the window open_ and close it when command ends
- ~~should NOT change 'modified'~~ (punting this for now; too risky)
- ~~ctrl-p/ctrl-n/up/down in `:` should not trigger 'inccommand' until user edits the command~~
    - going to wait for feedback before doing this

### Renamed `'incsubstitute'` to `'inccommand'` 

So that we can expand it to other commands besides `:substitute` without adding new options (that was already implied in the implementation, re `eap->is_live`).

### Future considerations

- block RPCs/events during incremental/live preview (`eap->is_live`)?
    - the buffer is in a weird state during the preview, don't want plugins interacting with it
    - done: #9783
- ~~instead of `u_undo_and_forget()`, it might be more robust to create a special/dummy buffer and throw it away~~
    - crazy talk: to show live previews of multi-buffer commands like `:argdo`, `:cdo`, overlay a new, temporary instance of `nvim`?

### CODE REVIEW

File: src/nvim/ex_cmds.c [hunk 1]
```diff
@@ -6189,24 +6179,32 @@ static void incsub_display(char_u *pat, char_u *sub,
...
   curbuf->b_p_ul = LONG_MAX;  // make sure we can undo all changes
   block_autocmds();   // disable events before incsub opening window/buffer
   emsg_off++;         // No error messages for live commands
 
-  do_sub(eap);
+  buf_T *incsub_buf = do_sub(eap);
 
   emsg_off--;
+  if (save_changedtick != curbuf->b_changedtick
+      && !u_undo_and_forget(1)) {
```
Comment (justinmk) [inline src/nvim/ex_cmds.c]: `u_undo_and_forget()` does not restore `b:changedtick`, which is used by some plugins like repeat.vim. @bfredl @KillTheMule should we restore it in `u_undo_and_forget()` or would it not be appropriate there?

**Comment by justinmk** [comment]
Only 2 tests failing, and valgrind seems OK.

**Comment by KillTheMule** [comment]
> add tests for feedkeys(':%s/foo/bar/'), feedkeys(":%s/foo/bar/\<CR>"), ...

If you could elaborate on this (I'm not sure what you mean), I could see if I can write them.

**Comment by justinmk** [comment]
@KillTheMule Generally, to verify that "live" substitution (or any "live" handling in general) should not be invoked during anything except actual interactive `:` use.

**Comment by KillTheMule** [comment]
Ok we have some basics: https://github.com/neovim/neovim/pull/5561/commits/561c1e39e05bac583f851de6165cd656058df7e1#diff-4dbe255ebfc94f6664c66ffeab750809R83

You mean along the lines of https://github.com/neovim/neovim/pull/5561/commits/561c1e39e05bac583f851de6165cd656058df7e1#diff-4dbe255ebfc94f6664c66ffeab750809R818, just with `incsubstute=''`? Note sure what to do about `any "live" handling in general` though.

**Comment by justinmk** [comment]
> Ok we have some basics: 561c1e3#diff-4dbe255ebfc94f6664c66ffeab750809R83

Confusingly, the `helpers.feed()` method is not equivalent to VimL `feedkeys()`. `helpers.feed()` simulates actual user input. Where I would like to verify that if plugins call `:substitute` or `feedkeys(':s...')`, incsub should not be invoked. Easy way to verify this is to check if the `incsub` buffer exists after either of those scenarios. (With this PR, the `incsub` buffer hangs around as an unlisted buffer.)

**Comment by justinmk** [comment]
Woo, ASAN looks good. Going to wrap this up now.

**Comment by KillTheMule** [comment]
> highlight bug #5226 (comment)

Note that this is a highlighting bug for single-line substitutions containing escaped chars, but makes the whole thing totally break down with multiline matches.

**Comment by justinmk** [comment]
Damn, that's disappointing. I would guess that multiline searches aren't common for many users, but it's still a bad bug.

**Comment by KillTheMule** [comment]
It still works as a substitution, just the preview is borked :) But it also means the data structure needs to be reworked, because right now it can only save one line that contains the whole match.

File: test/functional/ui/incsubstitute_spec.lua [hunk 2]
```diff
@@ -0,0 +1,1286 @@
...
+
+      execute(command)
+      expect([[
+        Inc substitution on
+        two LINES
+        ]])
+      end
+  end)
+
+  it('work then mappings move the cursor', function()
```
Comment (KillTheMule) [inline test/functional/ui/incsubstitute_spec.lua]: Small typo there: then -> when

**Comment by justinmk** [comment]
Renamed `'incsubstitute'` to `'inccommand'` so that we can expand it to other commands besides `:substitute` without adding new options (that was already implied in the implementation, re `eap->is_live`).

File: test/functional/ui/incsubstitute_spec.lua [hunk 3]
```diff
@@ -52,6 +52,31 @@ local function common_setup(screen, inccommand, text)
...
+describe(":substitute, inccommand=split does not trigger preview", function()
+  before_each(function()
+    clear()
+    common_setup(nil, "split", default_text)
+  end)
+
+  it("when invoked by feedkeys() in a script ", function()
+    source(':call feedkeys(":%s/tw/MO/g\\<CR>")')
+    wait()
+    eq(1, eval("bufnr('$')"))
```
Comment (KillTheMule) [inline test/functional/ui/incsubstitute_spec.lua]: Are you sure this checks what we want? By putting `\<CR>` at the end of the source, the command would finish, so the preview window is supposed to be closed at this point anyways. Maybe it's useful to check for it, but you can't be sure it didn't show inbetween,  can you?

**Comment by justinmk** [comment]
> Are you sure this checks what we want? By putting `<CR>` at the end of the source, the command would finish, so the preview window is supposed to be closed at this point anyways.

@KillTheMule The compromise made in this PR to avoid the special `buflist_new()` and `do_ecmd()` special-cases, is to keep the buffer around as an unlisted buffer. (That's more or less idiomatic, and less prone to bugs.)

Checking `bufnr('$')` is sufficient because `bufnr('$')` includes unlisted buffers. Note that before https://github.com/neovim/neovim/pull/5561/commits/48a986c030a8cb0578547f312b2b7d56dc73cf15 the test correctly fails.

**Comment by justinmk** [comment]
@bfredl @KillTheMule I added a [new test](https://github.com/neovim/neovim/pull/5561/commits/b0b8b277eb92634af6f65f5fd7585541a2e0813a#diff-74648f2c6258b7cd7be5e1d661b3a2ebR141) which compares `undotree()` before and after a canceled `:sub` preview. Here's the failure:

```
:substitute, 'inccommand' preserves undotree()
test/functional/ui/incsubstitute_spec.lua:164: Expected objects to be the same.
Passed in:
(table) {
 *[entries] = {
    [1] = {
      [seq] = 1
      [time] = 1478533735 }
    [2] = {
      [alt] = { ... more }
      [seq] = 3
      [time] = 1478533735 }
    [3] = {
      [seq] = 4
      [time] = 1478533735 }
   *[4] = {
     *[newhead] = 1
      [seq] = 5
      [time] = 1478533735 }
    [5] = {
      [curhead] = 1
      [seq] = 6
      [time] = 1478533735 }
    [6] = {
      [seq] = 7
      [time] = 1478533735 } }
  [save_cur] = 0
  [save_last] = 0
  [seq_cur] = 5
  [seq_last] = 7
  [synced] = 1
  [time_cur] = 1478533735 }
Expected:
(table) {
 *[entries] = {
    [1] = {
      [seq] = 1
      [time] = 1478533735 }
    [2] = {
      [alt] = { ... more }
      [seq] = 3
      [time] = 1478533735 }
    [3] = {
      [seq] = 4
      [time] = 1478533735 }
   *[4] = {
      [seq] = 5
      [time] = 1478533735 }
    [5] = {
      [curhead] = 1
      [seq] = 6
      [time] = 1478533735 }
    [6] = {
      [newhead] = 1
      [seq] = 7
      [time] = 1478533735 } }
  [save_cur] = 0
  [save_last] = 0
  [seq_cur] = 5
  [seq_last] = 7
  [synced] = 1
  [time_cur] = 1478533735 }
```

The difference is that `[newhead] = 1` on entry 6 (expected) vs entry entry 4 (actual). I don't understand `u_undo_and_forget()` at all, so I'm hoping you have some insight. Is this expected?

**Comment by KillTheMule** [comment]
I don't fully understand it either, but I think this is not expected. Could you rerun the test with `feed([[:%s/e<C-\><C-N>]])` (or if this does not fail, add keys til the first failure to produce a minimal example)? I'd have a shot at going through the logic then.

**Comment by bfredl** [comment]
Here is the thing. Whenever "curhead" points to a valid head, the value of "newhead" is meaningless (and really should be set to null). The undo state for a buffer is _logically_ the enum:
```
enum UndoState {
    CurrentHead(head),
    NewHead(head),
    EmptyTree
}
```
nvim _represents_ this as, whenever `curbuf->b_u_curhead` is nonnull it should be used as the current head, and `curbuf->b_u_newhead` is ignored.If the there is a current head, then this will be redoed on the next redo, and its parent will be undone on next undo.
Only if `b_u_curhead` is NULL, `b_u_newhead` will be used as the head to undo (and it is not possible to redo). Also both can be NULL, to indicate an empty undotree. (To be fair, this only strictly true when calling undo.c from the outside, in some places _within_ a function in undo.c both values might be meaningful) 

Apparently `undotree()` breaks this non-abstraction, this _cosmetic_ issue can easily be fixed by `ex_substitute` also saving and restoring `b_u_newhead`, but is doesn't reflect any error really how `u_undo_and_forget` manipulates the _actual_ state of the undo tree.

**Comment by justinmk** [comment]
Thank you @bfredl , see https://github.com/neovim/neovim/pull/5561/commits/e99890f3d6ea720dc25bab722978ab2f9af50d59

**Comment by justinmk** [comment]
Merging this, but I think the cursor movement should be revisited. The "jumping around" while building the `:sub` command is not really helpful. Especially with `inccommand=split` where results are shown without moving the cursor at all.

**Comment by Alok** [comment]
Could this be used to extend the `incsearch` option to highlight *all* candidate matches rather than just the current one (I think `emacs`does this)?

**Comment by justinmk** [comment]
@Alok that's being discussed on vim_dev so we'll wait for it to land there.

**Comment by gaving** [comment]
Congrats on the merge of this rockin' feature, this [and the associated PR's] are a great example of OSS done right!

**Comment by xdxter** [comment]
Thanks for the work! This feature **rocks**!!

Though, I wonder how hard would it be to make it work with the ```:Subvert``` command from tpope's [abolish](https://github.com/tpope/vim-abolish) plugin, which is immensely helpful in real-life substitution use cases.

**Comment by dsifford** [comment]
@justinmk 

> that's being discussed on vim_dev so we'll wait for it to land there.

You wouldn't happen to have a link to that discussion thread and/or an update on the status of that, would you?

Thanks in advance!

**Comment by justinmk** [comment]
@dsifford It's already implemented in Vim and Nvim 0.3: https://github.com/neovim/neovim/pull/7463

**Comment by dsifford** [comment]
@justinmk Is there a visual example of this somewhere I can take a look at? I'm thinking that maybe what I'm expecting is different than what's already implemented. I've got both `hlsearch` and `incsearch` enabled and it's still not doing what I'd expect.

**Comment by justinmk** [comment]
@dsifford Please always mention the exact version you're using. Also for questions about usage/configuration, try https://vi.stackexchange.com and the resources listed at https://neovim.io/community/.

**Comment by dsifford** [comment]
@justinmk Thanks for the info.

I think I may have found a bug in 0.2.2. (everything works as expected in vim proper, but not neovim -- even when running with `-u NONE`).

I'm gonna compile nightly here in a few and test and see if it's already been fixed before opening an issue on this.

**Comment by justinmk** [comment]
@dsifford I mentioned in https://github.com/neovim/neovim/pull/5561#issuecomment-383732551 the feature requires Nvim 0.3.

**Comment by dsifford** [comment]
D'oh! :confounded: 

I apologize for the noise. Thanks so much for your patience and kind feedback.

**Pull request merged**